### PR TITLE
Make FieldNamesFieldMapper responsible for adding its own doc fields

### PR DESCRIPTION
**Pull request by romseygeek**
The FieldNamesFieldMapper is a metadata mapper defining a field that
can be used for exists queries if a mapper does not use doc values or 
norms.  Currently, data is added to it via a special method on FieldMapper
that pulls the metadata mapper from a mapping lookup, checks to see
if it is enabled, and then adds the relevant value to a lucene document.

This is one of only two places that pulls a metadata mapper from the 
MappingLookup, and it would be nice to remove this method.  This commit
refactors field name handling by instead storing the names of fields to
index in the fieldnames field in a set on the ParseContext, and then 
building the field itself in FieldNamesFieldMapper.postParse().  This means
that all of the responsibility for enabling indexing, etc, is handled within
the metadata mapper itself.

This commit also removes indexing of all parent paths.  Previously, given
a field `parent.child.grandchild` we would index `parent`, `parent.child`
and `parent.child.grandchild` in the field names field, and this was originally
used to allow efficient exists queries on object fields.  However, since we
moved to using docvalue iterators for exists fields where possible, exists
queries against object fields have been built differently and so these
extra paths are unused.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java [hunk 1]
```diff
@@ -266,19 +264,6 @@ protected void indexScriptValues(SearchLookup searchLookup, LeafReaderContext re
         throw new UnsupportedOperationException("FieldMapper " + name() + " does not support [script]");
     }
 
-    protected final void createFieldNamesField(ParseContext context) {
-        assert fieldType().hasDocValues() == false : "_field_names should only be used when doc_values are turned off";
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:270]: is there a way to keep this assertion? I vaguely remember adding it while cleaning up some inconsistencies around using field_names where it would not make sense

File: server/src/main/java/org/elasticsearch/index/mapper/FieldNamesFieldMapper.java [hunk 2]
```diff
@@ -155,40 +156,14 @@ public FieldNamesFieldType fieldType() {
...
-                    }
-
-                    @Override
-                    public void remove() {
-                        throw new UnsupportedOperationException();
-                    }
-
-                };
-            }
-        };
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/FieldNamesFieldMapper.java:191]: on simplifying this and not indexing all the inner paths, that makes sense, but should we make that change in a separate PR?

File: server/src/main/java/org/elasticsearch/index/mapper/ParseContext.java [hunk 3]
```diff
@@ -490,6 +501,16 @@ public void addIgnoredField(String field) {
         public Collection<String> getIgnoredFields() {
             return Collections.unmodifiableCollection(ignoredFields);
         }
+
+        @Override
+        public void addFieldExistsField(String field) {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/ParseContext.java:506]: nit: I slightly prefer the previous name, shall we call it addFieldToFieldNames ?

File: server/src/main/java/org/elasticsearch/index/mapper/ParseContext.java [hunk 4]
```diff
@@ -490,6 +501,16 @@ public void addIgnoredField(String field) {
...
+            fieldNameFields.add(field);
+        }
+
+        @Override
+        public Collection<String> getFieldExistsFields() {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/ParseContext.java:511]: getFieldNames?

**Review by javanna — APPROVED** [review]
left a couple of small comments, LGTM otherwise. Thanks, this is a good cleanup.

File: server/src/main/java/org/elasticsearch/index/mapper/FieldNamesFieldMapper.java (hunk 2 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldNamesFieldMapper.java:191]: That's a good point, I'll pull this change into a separate issue.

File: server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java (hunk 1 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:270]: I've added an assertion in FieldNamesFieldMapper.postParse()

File: server/src/main/java/org/elasticsearch/index/mapper/ParseContext.java (hunk 3 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/ParseContext.java]: I ended up with `addtoFieldNames`
